2019/pkg/intcode/intcode.go

```diff
@@ -51,91 +66,67 @@ type Program struct {
-			p.instructions[dstIdx] = input
-		case opCodeWrite:
-			dstIdx := p.instructions[p.programCounter+1]
-			out := p.instructions[dstIdx]
-			if len(addressingModes) >= 1 && addressingModes[0] == immediateValue {
-				out = dstIdx
-			}
-			p.output = append(p.output, out)
-		case opCodeTerminate:
-			return p.instructions[0], nil
-		default:
-			return -1, fmt.Errorf("%w: %d", errInvalidOpCode, op)
+		if opcode == opCodeRead {
+			p.instructions[parameterIndexes[0]] = input
+		}
+		jumpTo, err := p.Exec(opcode, parameterIndexes)
+		if err != nil {
+			return err
 		}
 
-		p.programCounter += opCodeInstructionCount[op]
+		p.programCounter += opCodeInstructionCount[opcode]
+		if jumpTo != noJump {
+			p.programCounter = jumpTo
+		}
 	}
-	return p.instructions[0], nil
-}
-
-func (p *Program) Output() []int {
-	return p.output
 }
 
-func read2In1Out(programCounter int, input []int, addressingModes []addressingMode) (operand1, operand2, op int) {
-	operand1, operand2, output := input[programCounter+1], input[programCounter+2], input[programCounter+3]
-	if len(addressingModes) == 0 || addressingModes[0] == absoluteAddress {
-		operand1 = input[operand1]
+func (p *Program) Parse() (opCode, []int, error) {
+	opcode, modes, err := p.parseOpCodeAddressMode()
+	if err != nil {
+		return opCodeUnknown, []int{}, err
 	}
-	if len(addressingModes) <= 1 || addressingModes[1] == absoluteAddress {
-		operand2 = input[operand2]
-	}
-	return operand1, operand2, output
+	indexes := p.parseParameterIndexes(opcode, modes)
+	return opcode, indexes, err
 }
 
-// parseOpCodeAddressMode parses the opCode and the addressing modes from the input.
-// The opCode is a two-digit number based only on the ones and tens digit of the value.
-// The addressing mode is series of digits starting from the hundreths place.
-func parseOpCodeAddressMode(value int) (opCode, []addressingMode, error) {
-	opCodeValue := value
-	if value >= 100 {
+func (p *Program) parseOpCodeAddressMode() (opCode, []addressingMode, error) {
+	// Parse opcode converts the instruction to an opcode.
+	instruction := p.instructions[p.programCounter]
+	opCodeValue := instruction
+	if opCodeValue >= 100 {
 		opCodeValue %= 100
 	}
-
 	if ok := isValid[opCode(opCodeValue)]; !ok {
-		return opCodeUnknown, []addressingMode{}, fmt.Errorf("%w: %d", errInvalidOpCode, opCodeValue)
+		return opCodeUnknown, nil, fmt.Errorf("%w: opCode=%d", errInvalidOpCode, opCodeValue)
 	}
-	return opCode(opCodeValue), parseAddressingModes(value / 100), nil
-}
 
-// parseAddressingModes converts the input integer into a list of its digits,
-// then for each digit, converts it to the corresponding addressingMode.
-func parseAddressingModes(modes int) []addressingMode {
-	digits := intToDigits(modes)
+	// Parse address modes converts the input integer into a list of its digits.
+	// Each digit is converted it to the corresponding addressingMode.
+	modeValue := instruction / 100
+	digits := intToDigits(modeValue)
```

Comment:
Note about performance, you are creating a slice for each `intToDigits` call.

Reply:
what are alternatives to avoid this?  I could remove the method `intToDigits` and execute that code inline where its used.

Reply:
Yes, one is to execute it inline. This is the "safest" with regards to performance overhead, but can get messy if it would need a lot of code.

Then another option is to write a custom iterator, something like:
```
digit := iterateDigits(modeValue)
for digit.Next() {
    fmt.Println(digit.Value())
}
````

The other is using callbacks:

```
iterateDigits(modeValue, func(digit int) { ... })
```

Note, the usefulness of callback approach will depend on whether the compiler is able to inline the `iterateDigits` func or not.

Reply:
nice. i like the iterator, will add that.